CE-922 - Update "Automated upgrades" (#1461)
hashicorp/web-unified-docs: content/consul/v1.22.x/content/docs/upgrade/automated.mdx

---
layout: docs
page_title: Automated Upgrades
description: >-
  Automated Upgrades simplify the process for updating Consul. Learn how Consul can gracefully transition from existing server agents to a new set of server agents without Consul downtime.
---

# Automated Upgrades

One of the fundamental operations to be performed in any production environment is to upgrade to a new version without experiencing downtime and to minimize any risk factors that could lead to future outages.

Consul Enterprise allows to automatically upgrade a cluster of Consul servers to a new version by adding updated server nodes to the cluster.
Consul monitors the amount of servers in the cluster and, when a sufficient amount of new server nodes are joined, Consul promotes them to voters, elects a new leader among the new servers, and demotes the old version servers to non-voters. Once this demotion occurs, the old servers can be safely removed from the cluster.

@include 'alerts/enterprise-alert.mdx'

> The functionality can be used in different use-case scenarios:

Reviewer: ```suggestion
Automated upgrades can help you in the following scenarios:
```

1. Upgrade to a new Consul version - This scenario is useful when you want to upgrade Consul servers to a newer version without downtime.
1. Migrate to different server nodes - This scenario is useful when you have a new Consul server image with security patches and want to replace the running Consul server instances automatically, without experiencing downtime.

## Prerequisites

To test the automated upgrades feature explained in this tutorial you will need:

- A Consul Enterprise datacenter with three servers running Consul Enterprise.
- Three extra nodes with the latest Consul Enterprise binary installed to replace the old servers.

The different Consul versions need to be compatible for a direct upgrade. Check [Protocol Compatibility Promise](/consul/docs/upgrade/compatibility) and [Upgrade instructions](/consul/docs/upgrade/instructions) to make sure the versions you are using are compatible.

> ## Upgrade to a new Consul version

Reviewer: ```suggestion
## Workflow
```

To upgrade Consul servers to a newer version the following steps are required:
1. Verify the Automated Upgrades feature is enabled
1. Add new Consul servers
1. Verify upgrade

> 1. Stop old servers

Reviewer: ```suggestion

1. Verify that automated upgrades is enabled
1. Add new Consul servers
1. Verify upgrade
1. Stop old servers
```

### Verify the Automated Upgrades feature is enabled

The automated upgrades feature is enabled by default in Consul Enterprise. 

> Verify the configuration for your datacenter by using the `consul operator autopilot` command.

Reviewer: ```suggestion
To verify the configuration in your datacenter, use the `consul operator autopilot` command.
```

```shell-session
$ consul operator autopilot get-config
```

The command returns the whole configuration for Autopilot features, check `DisableUpgradeMigration` is set to `false`.

<CodeBlockConfig hideClipboard highlight="8">

```plaintext
consul operator autopilot get-config
CleanupDeadServers = true
LastContactThreshold = 200ms
MaxTrailingLogs = 250
MinQuorum = 0
ServerStabilizationTime = 10s
RedundancyZoneTag = ""
DisableUpgradeMigration = false
UpgradeVersionTag = ""
```

</CodeBlockConfig>

If the feature is disabled in your datacenter, enable it with the `consul operator autopilot set-config -disable-upgrade-migration=false` command.

### Add new servers

> In this example, we used a Consul Enterprise `1.21.7` datacenter with three servers and `bootstrap_expect = 3`.

Reviewer: ```suggestion
The following example uses an existing Consul Enterprise `1.21.7` datacenter with three servers and `bootstrap_expect = 3`. 
```

Use `consul members` to verify initial state of the cluster.

```shell-session
$ consul members
Node             Address           Status  Type    Build       Protocol  DC   Partition  Segment
consul-server-1  172.18.0.12:8301  alive   server  1.21.7+ent  2         dc1  default    <all>
consul-server-2  172.18.0.8:8301   alive   server  1.21.7+ent  2         dc1  default    <all>
consul-server-3  172.18.0.6:8301   alive   server  1.21.7+ent  2         dc1  default    <all>
```

Old server nodes are named using the `consul-server-<number>` scheme while the new server nodes will use `consul-server-<number>` scheme.

> 

Author: In case the new servers have a newer version of Consul there is no specific configuration required.  The other case, when you want to make the "upgrade" using the `meta` parameter, requires some config that is shown.

consul-server-1   630ffcac-8f68-e9e0-5cca-54c3fa5f7290  172.18.0.12:8300  leader    true   3             85            -
consul-server-2   bffdeec5-27f0-2116-99b1-68714679c2a0  172.18.0.8:8300   follower  true   3             85            0 commits
consul-server-3   4aedd7d4-d5e4-109e-33dc-f6c2a2fd98c9  172.18.0.6:8300   follower  true   3             85            0 commits
consul-server-1a  16a4ffeb-a8d4-b42d-edf1-cdec993fdcdc  172.18.0.17:8300  follower  false  3             85            0 commits
consul-server-2a  73e267e5-f817-0577-bbe5-2f843adff357  172.18.0.4:8300   follower  false  3             85            0 commits
```

Until enough servers with the new Consul servers are part of the datacenter, Consul will keep the added servers as `non-voter` members. When the number of new servers is enough to form a quorum and elect a new leader, autopilot will promote the new servers to voters, trigger a new leader election, and finally demote the old servers.

## Verify upgrade

Once the third new server is started, Consul autopilot starts the upgrade process.

> New nodes are first promoted to `voter` state.

Reviewer: ```suggestion
First, new nodes are promoted to `voter` state.
```

Once the upgrade is rolled out, remove the old servers from the datacenter using the `consul leave` command. 

## Migrate to different server nodes

In some environments, Consul is not the only component that needs to be upgraded to ensure security or stability for the node. For example, you might
need to roll out node updates to apply OS patches or new configuration settings.

To cover these cases, Consul autopilot offers the possibility to set the `UpgradeVersionTag` parameter to override the parameter that Consul uses for Automated Upgrades. This logic can be used for updating the datacenter nodes even without the need to change Consul version.

When `UpgradeVersionTag` is configured, Consul will use its value to look for a version in each server's specified [`node_meta`](/consul/docs/reference/agent/configuration-file/node#node_meta) tag.

For example, if `UpgradeVersionTag` is set to `build`, and `node-meta.build` is set to `0.0.2` in server configuration, that server's version will be
`0.0.2` when considered by autopilot for a migration. The upgrade logic will follow semantic versioning and the version string must be in the form of either `X`, `X.Y`, or `X.Y.Z`.

> To upgrade Consul servers using `UpgradeVersionTag` the following steps are required:

Reviewer: ```suggestion
To upgrade Consul servers using `UpgradeVersionTag`, complete the following steps:
```

1. Configure autopilot for existing servers
1. Add new Consul servers 
1. Verify upgrade
1. Stop old servers

> ## Configure autopilot for existing servers

Reviewer: ```suggestion
### Configure autopilot for existing servers
```

> For this example, we use as prerequisites the three servers started in the previous example, `server-1a`, `server-2a`, and `server-3a`. To perform the migration you can start three extra servers, `server-1b`, `server-2b`, and `server-3b`, or, alternatively reuse the ones used earlier as prerequisites as long as the Consul binary on the servers is changed to match the one running on `server-1a`, `server-2a` and `server-3a`.

Reviewer: ```suggestion
The following example begins with three servers: `server-1a`, `server-2a`, and `server-3a`. Three additional servers are used, names `server-1b`, `server-2b`, amd `server-3b`.
```

</CodeTabs>

> Once the third new server is started, autopilot detects the possibility to reach a quorum among the new servers and promotes them as voters, triggers a new leader election, and demotes the old nodes as non-voters. This time the version comparison happens not using the Consul internal version but the value expressed in the `build` value.

Reviewer: ```suggestion
After the third new server starts, autopilot detects the possibility of a quorum among the new servers, and promotes them to voters.
```

```

Finally, when a new leader is elected among the new Consul servers, the old leader is demoted to `non-voter`.

```shell-session
$ consul operator raft list-peersrs
Node              ID                                    Address           State     Voter  RaftProtocol  Commit Index   Trails Leader By
consul-server-1a  16a4ffeb-a8d4-b42d-edf1-cdec993fdcdc  172.18.0.17:8300  follower  false  3             157            0 commits
consul-server-2a  73e267e5-f817-0577-bbe5-2f843adff357  172.18.0.4:8300   follower  false  3             157            0 commits
consul-server-3a  626faa96-91af-53b1-da97-21435d4c5f62  172.18.0.15:8300  follower  false  3             157            0 commits
consul-server-1b  c04510f4-4d6d-9b54-3ac8-8843b848b24a  172.18.0.9:8300   leader    true   3             157            0 commits
consul-server-2b  7ba9deda-fd46-cb10-6341-b32eb2db87de  172.18.0.6:8300   follower  true   3             157            0 commits
consul-server-3b  3908d249-3da3-2262-7405-da5b80337716  172.18.0.11:8300  follower  true   3             157            0 commits
```

> The process can also be observed from the leader's logs:

Reviewer: ```suggestion
You can also observe this process in the leader's logs:
```

```log
## ...
[INFO]  agent.server.autopilot: Promoting server: id=c04510f4-4d6d-9b54-3ac8-8843b848b24a address=172.18.0.9:8300 name=consul-server-1b
## ...
[INFO]  agent.server.autopilot: Promoting server: id=7ba9deda-fd46-cb10-6341-b32eb2db87de address=172.18.0.6:8300 name=consul-server-2b
## ...
[INFO]  agent.server.autopilot: Promoting server: id=3908d249-3da3-2262-7405-da5b80337716 address=172.18.0.11:8300 name=consul-server-3b
## ...
[INFO]  agent.server.autopilot: Demoting server: id=626faa96-91af-53b1-da97-21435d4c5f62 address=172.18.0.15:8300 name=consul-server-3a
## ...
[INFO]  agent.server.autopilot: Demoting server: id=73e267e5-f817-0577-bbe5-2f843adff357 address=172.18.0.4:8300 name=consul-server-2a
## ...
[INFO]  agent.server.autopilot: Transferring leadership to new server: id=c04510f4-4d6d-9b54-3ac8-8843b848b24a address=172.18.0.9:8300
## ...
[INFO]  agent.server.raft: entering follower state: follower="Node at 172.18.0.17:8300 [Follower]" leader-address= leader-id=
## ...
[INFO]  agent.server: cluster leadership lost
## ...
[INFO]  agent.server: New leader elected: payload=consul-server-1b
```

</CodeBlockConfig>

### Stop old servers

> Once the upgrade is rolled out, remove the old servers from the datacenter using the `consul leave` command. 

Reviewer: ```suggestion
Once the upgrade is rolled out, use the `consul leave ` command to remove the old servers from the datacenter. 
```

## Next steps

In this document, you learned how to upgrade your Consul datacenter by using autopilot's Automated Upgrades functionality. 

> Use the feature when planning upgrades to reduce manual steps to a minimum.

Reviewer: ```suggestion
```
